File: src/feed.rs
use crate::config::{FeedType, Source, SourceType};

#[derive(Debug, Clone)]
pub struct Article {
    pub title: String,
    pub link: String,
    pub description: String,
    pub pub_date: String,
    pub author: String,
}

impl Article {
    pub fn summary(&self) -> String {
        let plain = html2text::from_read(self.description.as_bytes(), 80);
        // Trim excessive whitespace/newlines
        plain
            .lines()
            .map(|l| l.trim())
            .filter(|l| !l.is_empty())
            .collect::<Vec<_>>()
            .join("\n")
    }
}

#[derive(Debug, Clone)]
pub enum FetchError {
    Network(String),
    Parse(String),
}

impl std::fmt::Display for FetchError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            FetchError::Network(e) => write!(f, "Network error: {}", e),
            FetchError::Parse(e) => write!(f, "Parse error: {}", e),
        }
    }
}

pub fn fetch_feed(source: &Source) -> Result<Vec<Article>, FetchError> {
    match &source.source_type {
        SourceType::Feed { url, feed_type } => fetch_rss_atom(url, feed_type),
        SourceType::Irc { .. } => {
            // IRC sources are not yet supported for fetching
            Err(FetchError::Network(
                "IRC source fetching is not yet implemented".to_string(),
            ))
        }
    }
}

fn fetch_rss_atom(url: &str, feed_type: &FeedType) -> Result<Vec<Article>, FetchError> {
    // Build a client with proper configuration for feed fetching
    let client = reqwest::blocking::Client::builder()
        .user_agent("Mozilla/5.0 (compatible; termnews/0.1; +https://github.com/adamdb/newsterm)")
        .timeout(std::time::Duration::from_secs(30))
        .redirect(reqwest::redirect::Policy::limited(10))
        .build()
        .map_err(|e| FetchError::Network(format!("Failed to build HTTP client: {}", e)))?;

    let response = client
        .get(url)
        .send()
        .map_err(|e| FetchError::Network(e.to_string()))?;

    let body = response
        .text()
        .map_err(|e| FetchError::Network(e.to_string()))?;

    match feed_type {
        FeedType::Rss => parse_rss(&body),
        FeedType::Atom => parse_atom(&body),
    }
}

fn parse_rss(body: &str) -> Result<Vec<Article>, FetchError> {
    let channel = body
        .parse::<rss::Channel>()
        .map_err(|e| FetchError::Parse(e.to_string()))?;

    let articles = channel
        .items()
        .iter()
        .map(|item| Article {
            title: item.title().unwrap_or("(no title)").to_string(),
            link: item.link().unwrap_or("").to_string(),
            description: item.description().unwrap_or("").to_string(),
            pub_date: item.pub_date().unwrap_or("").to_string(),
            author: item
                .author()
                .or_else(|| {
                    item.dublin_core_ext()
                        .and_then(|dc| dc.creators().first().map(|s| s.as_str()))
                })
                .unwrap_or("")
                .to_string(),
        })
        .collect();

    Ok(articles)
}

fn parse_atom(body: &str) -> Result<Vec<Article>, FetchError> {
    let feed: atom_syndication::Feed = body
        .parse()
        .map_err(|e: atom_syndication::Error| FetchError::Parse(e.to_string()))?;

    let articles = feed
        .entries()
        .iter()
        .map(|entry| {
            let description = entry
                .content()
                .and_then(|c| c.value())
                .map(|s| s.to_string())
                .or_else(|| entry.summary().map(|s| s.value.clone()))
                .unwrap_or_default();

            let link = entry
                .links()
                .first()
                .map(|l| l.href())
                .unwrap_or("")
                .to_string();

            let pub_date = entry
                .published()
                .map(|d| d.to_rfc3339())
                .unwrap_or_else(|| entry.updated().to_rfc3339());

            let author = entry
                .authors()
                .first()
                .map(|a| a.name())
                .unwrap_or("")
                .to_string();

            Article {
                title: entry.title().value.to_string(),
                link,
                description,
                pub_date,
                author,
            }
        })
        .collect();

    Ok(articles)
}

File: internal/feed/feed.go
// Package feed handles fetching and parsing RSS and Atom feeds.
package feed

import (
	"context"
	"fmt"
	"html"
	"net/http"
	"regexp"
	"strings"
	"time"

	"github.com/adamdb/termnews/internal/config"
	"github.com/mmcdole/gofeed"
)

// Article represents a single news article.
type Article struct {
	Title       string
	Link        string
	Description string
	PubDate     time.Time
	Author      string
	Categories  []string // Categories/tags for the article
	Read        bool     // Whether the article has been read
	Bookmarked  bool     // Whether the article is bookmarked
	GUID        string   // Unique identifier for tracking read status
}

// Summary returns a plain text summary of the article description.
func (a *Article) Summary() string {
	return htmlToPlainText(a.Description)
}

// FormattedDate returns a human-readable date string.
func (a *Article) FormattedDate() string {
	if a.PubDate.IsZero() {
		return ""
	}
	now := time.Now()
	diff := now.Sub(a.PubDate)

	switch {
	case diff < time.Hour:
		mins := int(diff.Minutes())
		if mins <= 1 {
			return "just now"
		}
		return fmt.Sprintf("%dm ago", mins)
	case diff < 24*time.Hour:
		hours := int(diff.Hours())
		if hours == 1 {
			return "1h ago"
		}
		return fmt.Sprintf("%dh ago", hours)
	case diff < 7*24*time.Hour:
		days := int(diff.Hours() / 24)
		if days == 1 {
			return "1d ago"
		}
		return fmt.Sprintf("%dd ago", days)
	default:
		return a.PubDate.Format("Jan 2, 2006")
	}
}

// FetchError represents an error that occurred during feed fetching.
type FetchError struct {
	Type    string // "network" or "parse"
	Message string
}

func (e FetchError) Error() string {
	return fmt.Sprintf("%s error: %s", e.Type, e.Message)
}

// FetchResult holds the result of fetching a feed.
type FetchResult struct {
	SourceIndex int
	Articles    []Article
	Error       error
}

// Fetch fetches articles from a feed source.
func Fetch(ctx context.Context, source *config.Source) ([]Article, error) {
	client := &http.Client{
		Timeout: 30 * time.Second,
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, source.URL, nil)
	if err != nil {
		return nil, FetchError{Type: "network", Message: err.Error()}
	}

	req.Header.Set("User-Agent", "Mozilla/5.0 (compatible; termnews/1.0; +https://github.com/adamdb/termnews)")
	req.Header.Set("Accept", "application/rss+xml, application/atom+xml, application/xml, text/xml")

	resp, err := client.Do(req)
	if err != nil {
		return nil, FetchError{Type: "network", Message: err.Error()}
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, FetchError{Type: "network", Message: fmt.Sprintf("HTTP %d: %s", resp.StatusCode, resp.Status)}
	}

	parser := gofeed.NewParser()
	feed, err := parser.Parse(resp.Body)
	if err != nil {
		return nil, FetchError{Type: "parse", Message: err.Error()}
	}

	articles := make([]Article, 0, len(feed.Items))
	for _, item := range feed.Items {
		article := Article{
			Title:       strings.TrimSpace(item.Title),
			Link:        strings.TrimSpace(item.Link),
			Description: item.Description,
			Categories:  item.Categories,
		}

		// Set publication date
		if item.PublishedParsed != nil {
			article.PubDate = *item.PublishedParsed
		} else if item.UpdatedParsed != nil {
			article.PubDate = *item.UpdatedParsed
		}

		// Set author
		if item.Author != nil {
			article.Author = strings.TrimSpace(item.Author.Name)
		} else if len(item.Authors) > 0 && item.Authors[0] != nil {
			article.Author = strings.TrimSpace(item.Authors[0].Name)
		}

		// Use content if description is empty
		if article.Description == "" && item.Content != "" {
			article.Description = item.Content
		}

		// Set GUID for tracking
		if item.GUID != "" {
			article.GUID = item.GUID
		} else {
			article.GUID = item.Link
		}

		articles = append(articles, article)
	}

	return articles, nil
}

// FetchAsync fetches a feed asynchronously and sends the result to a channel.
func FetchAsync(ctx context.Context, source *config.Source, index int, results chan<- FetchResult) {
	articles, err := Fetch(ctx, source)
	results <- FetchResult{
		SourceIndex: index,
		Articles:    articles,
		Error:       err,
	}
}

// htmlToPlainText converts HTML to plain text.
func htmlToPlainText(htmlStr string) string {
	// Unescape HTML entities
	text := html.UnescapeString(htmlStr)

	// Remove script and style elements
	reScript := regexp.MustCompile(`(?is)<script[^>]*>.*?</script>`)
	text = reScript.ReplaceAllString(text, "")

	reStyle := regexp.MustCompile(`(?is)<style[^>]*>.*?</style>`)
	text = reStyle.ReplaceAllString(text, "")

	// Replace common block elements with newlines
	blockTags := []string{"br", "p", "div", "li", "tr", "h1", "h2", "h3", "h4", "h5", "h6"}
	for _, tag := range blockTags {
		reTag := regexp.MustCompile(`(?i)<` + tag + `[^>]*>`)
		text = reTag.ReplaceAllString(text, "\n")
		reTagClose := regexp.MustCompile(`(?i)</` + tag + `>`)
		text = reTagClose.ReplaceAllString(text, "\n")
	}

	// Remove all other HTML tags
	reTags := regexp.MustCompile(`<[^>]+>`)
	text = reTags.ReplaceAllString(text, "")

	// Clean up whitespace
	lines := strings.Split(text, "\n")
	var cleanLines []string
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line != "" {
			cleanLines = append(cleanLines, line)
		}
	}

	return strings.Join(cleanLines, "\n")
}
